hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/heartbeat/HoodieHeartbeatClient.java

```diff
@@ -0,0 +1,196 @@
+import java.io.Serializable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+/**
+ * This class creates heartbeat for hudi client. This heartbeat is used to ascertain whether the running job is or not.
+ * NOTE: Due to CPU contention on the driver/client node, the heartbeats could be delayed, hence it's important to set
+ *       the value high enough to avoid that possibility.
+ */
+@NotThreadSafe
+public class HoodieHeartbeatClient implements AutoCloseable, Serializable {
+
+  // TODO : Throw new exception from thread
+  private static final Logger LOG = LogManager.getLogger(HoodieHeartbeatClient.class);
+
+  // path to the hearbeat folder where all writers are updating their heartbeats
+  private final transient FileSystem fs;
+  private final String basePath;
+  private String heartBeatFolderPath;
+  private String currentInstantTime;
+  private int numHeartBeatsForCurrentInstantTime;
+  // heartbeat interval in millis
+  private final long heartbeatIntervalInMillis;
+  private final transient ExecutorService executorService = Executors.newSingleThreadExecutor();
+  private volatile boolean shutdownRequested;
+  private boolean isHearbeatStarted = false;
+  private boolean isHeatBeatStopped = false;
+  private transient Future<?> heartBeatFuture;
+  private Long lastHeartBeatTime;
+  // This is required for testing. Problem : if we set heartbeatIntervalInSeconds really large, test takes longer, if
+  // we set it small and you are debugging with breakpoint, causes issues. Need to implement a Mock in tests
+  private final Boolean skipHeartBeatCheck;
+
+  public HoodieHeartbeatClient(FileSystem fs, String basePath, long heartbeatIntervalInSeconds) {
+    ValidationUtils.checkArgument(heartbeatIntervalInSeconds >= 1, "Cannot set heartbeat lower than 1 second");
+    this.fs = fs;
+    this.basePath = basePath;
+    this.heartBeatFolderPath = HoodieTableMetaClient.getHeartbeatFolderPath(basePath);
+    this.heartbeatIntervalInMillis = heartbeatIntervalInSeconds * 1000L;
+    this.skipHeartBeatCheck = true;
+  }
+
+  public void start(String instantTime) {
+    LOG.info("Received request to start heartbeat for instant time " + instantTime);
+    if (isHearbeatStarted && instantTime == currentInstantTime) {
+      // heartbeat already started, NO_OP
+      return;
+    } else if (instantTime != currentInstantTime) {
+      if (this.currentInstantTime != null) {
+        LOG.warn("Stopping heartbeat for previous instant time " + this.currentInstantTime);
+        this.stop(this.currentInstantTime);
+      }
+      this.currentInstantTime = instantTime;
+      this.numHeartBeatsForCurrentInstantTime = 0;
+      this.shutdownRequested = false;
+    }
+    try {
+      updateHeartbeat(instantTime);
+    } catch (Exception e) {
+      throw new RuntimeException("Unable to generate heartbeat");
+    }
+    this.heartBeatFuture = executorService.submit(() -> {
```

Comment:
can we use a TimerTask or TimerService. it can also support multiple of these schedules

Reply:
Added.